src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +85,581 @@ impl ReductionPushdown {
+
+        // skip reduction pushdown if there are no eligible aggregations to push
+        // down
+        if single_input_aggs_to_push.iter().all(|aggs| aggs.is_empty())
+            && no_input_aggs_to_push.is_empty()
+        {
+            return None;
+        }
+
+        let join_graph = create_join_graph(&old_join_mapper, equivalences)?;
+
+        Some(Self {
+            new_inputs: inputs.to_vec(),
+            outer_group_key: group_key,
+            outer_aggs,
+            downcast_partial_reduce: Vec::new(),
+            next_new_join_column: 0,
+            permutation: HashMap::new(),
+            evaluated_expressions: HashMap::new(),
+            single_input_aggs_to_push,
+            no_input_aggs_to_push,
+            is_delta: matches!(implementation, expr::JoinImplementation::DeltaQuery(_)),
+            old_input_keys: input_types
+                .iter()
+                .map(|typ| typ.keys.clone())
+                .collect::<Vec<_>>(),
+            equivalences,
+            aggs_not_pushed_support,
+            old_join_mapper,
+            join_graph,
+            pushdown_succeeded: false,
+        })
+    }
+
+    /// Construct the reduction pushdown, if able
+    fn get_relation_with_reduction_pushed(
+        mut self,
+        monotonic: &bool,
+        expected_group_size: &Option<usize>,
+    ) -> Option<MirRelationExpr> {
+        // Attempt reduction pushdown.
+        // Try two ways to construct a new join where at least one of the
+        // inputs gains a new inner reduce:
+        // 1) if there are aggregations on a single input, try to push them
+        // down. No-input aggregations will be pushed down to
+        // the first input where we can successfully push down single-input
+        // aggregations.
+        if self
+            .single_input_aggs_to_push
+            .iter()
+            .any(|aggs| !aggs.is_empty())
+        {
+            for idx in 0..self.new_inputs.len() {
+                self.try_pushdown(idx, true);
+            }
+        }
+        // 2) If no single-input aggregations get pushed down, then try pushing
+        // down no-input aggregations to any input
+        if !self.pushdown_succeeded && !self.no_input_aggs_to_push.is_empty() {
+            // reset the mapping of the old relation to the new relation.
+            // this is ok because none of the join inputs should have changed,
+            // so the mapping should be the identity mapping.
+            self.permutation = HashMap::new();
+            self.next_new_join_column = 0;
+            self.evaluated_expressions = HashMap::new();
+
+            for idx in 0..self.new_inputs.len() {
+                self.try_pushdown(idx, false);
+            }
+        }
+
+        if self.pushdown_succeeded {
+            // construct the new join
+            let mut new_inputs = Vec::new();
+            std::mem::swap(&mut new_inputs, &mut self.new_inputs);
+            let new_equivalences = self
```

Comment:
There is a panic on this line:

```
2021-02-03T13:07:53.577817Z ERROR panic: <unnamed>: no entry found for key
   0: materialized::handle_panic
             at wangandi-reduceaggs/src/materialized/src/bin/materialized/main.rs:568:65
   1: core::ops::function::Fn::call
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:581:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:484:9
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:153:18
   5: rust_begin_unwind
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:483:5
   6: core::panicking::panic_fmt
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/panicking.rs:85:14
   7: core::option::expect_failed
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/option.rs:1226:5
   8: core::option::Option<T>::expect
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/option.rs:346:21
      <std::collections::hash::map::HashMap<K,V,S> as core::ops::index::Index<&Q>>::index
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/collections/hash/map.rs:1095:9
      expr::scalar::MirScalarExpr::permute_map::{{closure}}
             at wangandi-reduceaggs/src/expr/src/scalar/mod.rs:206:26
      expr::scalar::MirScalarExpr::visit_mut
             at wangandi-reduceaggs/src/expr/src/scalar/mod.rs:182:9
   9: expr::scalar::MirScalarExpr::permute_map
             at wangandi-reduceaggs/src/expr/src/scalar/mod.rs:204:9
  10: transform::reduction_pushdown::ReductionPusher::get_relation_with_reduction_pushed::{{closure}}::{{closure}}
             at wangandi-reduceaggs/src/transform/src/reduction_pushdown.rs:380:33
      core::iter::adapters::map_fold::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/adapters/mod.rs:905:28
      core::iter::traits::iterator::Iterator::fold
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/traits/iterator.rs:2023:21
      <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/adapters/mod.rs:945:9
  11: core::iter::traits::iterator::Iterator::for_each
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/traits/iterator.rs:678:9
      <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::spec_extend
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/vec.rs:2378:17
      <alloc::vec::Vec<T> as alloc::vec::SpecFromIterNested<T,I>>::from_iter
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/vec.rs:2152:9
      <alloc::vec::Vec<T> as alloc::vec::SpecFromIter<T,I>>::from_iter
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/vec.rs:2162:9
      <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/vec.rs:2002:9
      core::iter::traits::iterator::Iterator::collect
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/traits/iterator.rs:1670:9
      transform::reduction_pushdown::ReductionPusher::get_relation_with_reduction_pushed::{{closure}}
             at wangandi-reduceaggs/src/transform/src/reduction_pushdown.rs:373:21
      core::iter::adapters::map_fold::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/adapters/mod.rs:905:28
      core::iter::traits::iterator::Iterator::fold
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/traits/iterator.rs:2023:21
      <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/adapters/mod.rs:945:9
  12: core::iter::traits::iterator::Iterator::for_each
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/traits/iterator.rs:678:9
      <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::spec_extend
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/vec.rs:2378:17
      <alloc::vec::Vec<T> as alloc::vec::SpecFromIterNested<T,I>>::from_iter
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/vec.rs:2152:9
      <alloc::vec::Vec<T> as alloc::vec::SpecFromIter<T,I>>::from_iter
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/vec.rs:2162:9
      <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/vec.rs:2002:9
      core::iter::traits::iterator::Iterator::collect
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/iter/traits/iterator.rs:1670:9
      transform::reduction_pushdown::ReductionPusher::get_relation_with_reduction_pushed
             at wangandi-reduceaggs/src/transform/src/reduction_pushdown.rs:369:36
  13: transform::reduction_pushdown::ReductionPushdown::action
             at wangandi-reduceaggs/src/transform/src/reduction_pushdown.rs:107:54
  14: <transform::reduction_pushdown::ReductionPushdown as transform::Transform>::transform::{{closure}}
             at wangandi-reduceaggs/src/transform/src/reduction_pushdown.rs:32:13
      expr::relation::MirRelationExpr::visit_mut_pre
             at wangandi-reduceaggs/src/expr/src/relation/mod.rs:904:9
  15: expr::relation::MirRelationExpr::visit_mut_pre::{{closure}}
             at wangandi-reduceaggs/src/expr/src/relation/mod.rs:905:29
      expr::relation::MirRelationExpr::visit1_mut::{{closure}}
             at wangandi-reduceaggs/src/expr/src/relation/mod.rs:866:13
      expr::relation::MirRelationExpr::try_visit1_mut
             at wangandi-reduceaggs/src/expr/src/relation/mod.rs:819:17
      expr::relation::MirRelationExpr::visit1_mut
             at wangandi-reduceaggs/src/expr/src/relation/mod.rs:865:9
      expr::relation::MirRelationExpr::visit_mut_pre
             at wangandi-reduceaggs/src/expr/src/relation/mod.rs:905:9
  16: <transform::reduction_pushdown::ReductionPushdown as transform::Transform>::transform
             at wangandi-reduceaggs/src/transform/src/reduction_pushdown.rs:31:9
  17: transform::Optimizer::transform
             at wangandi-reduceaggs/src/transform/src/lib.rs:178:13
      transform::dataflow::inline_views
             at wangandi-reduceaggs/src/transform/src/dataflow.rs:166:9
      transform::dataflow::optimize_dataflow
             at wangandi-reduceaggs/src/transform/src/dataflow.rs:27:5
  18: coord::coord::Coordinator<C>::ship_dataflow::{{closure}}
             at wangandi-reduceaggs/src/coord/src/coord.rs:3172:9
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
  19: coord::coord::Coordinator<C>::sequence_peek::{{closure}}
             at wangandi-reduceaggs/src/coord/src/coord.rs:2301:17
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
  20: coord::coord::Coordinator<C>::sequence_plan::{{closure}}
             at wangandi-reduceaggs/src/coord/src/coord.rs:1571:17
  21: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
      coord::coord::Coordinator<C>::message_statement_ready::{{closure}}
             at wangandi-reduceaggs/src/coord/src/coord.rs:499:17
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
      coord::coord::Coordinator<C>::serve::{{closure}}
             at wangandi-reduceaggs/src/coord/src/coord.rs:401:21
      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/future/mod.rs:80:19
  22: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/park/thread.rs:263:54
      tokio::coop::with_budget::{{closure}}
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/coop.rs:121:9
      std::thread::local::LocalKey<T>::try_with
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/local.rs:272:16
      std::thread::local::LocalKey<T>::with
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/local.rs:248:9
      tokio::coop::with_budget
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/coop.rs:114:5
      tokio::coop::budget
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/coop.rs:98:5
      tokio::park::thread::CachedParkThread::block_on
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/park/thread.rs:263:31
  23: tokio::runtime::enter::Enter::block_on
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/runtime/enter.rs:151:13
      tokio::runtime::thread_pool::ThreadPool::block_on
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/runtime/thread_pool/mod.rs:71:9
  24: tokio::runtime::Runtime::block_on
             at .cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.1.0/src/runtime/mod.rs:452:43
  25: coord::coord::serve::{{closure}}::{{closure}}
             at wangandi-reduceaggs/src/coord/src/coord.rs:3352:31
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys_common/backtrace.rs:137:18
  26: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/mod.rs:464:17
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panic.rs:308:9
      std::panicking::try::do_call
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:381:40
      std::panicking::try
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:345:19
      std::panic::catch_unwind
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panic.rs:382:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/thread/mod.rs:463:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:227:5
  27: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/boxed.rs:1042:9
      <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/alloc/src/boxed.rs:1042:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/sys/unix/thread.rs:87:17
  28: start_thread
  29: clone
````
To reproduce:
```
DROP SCHEMA public CASCADE;
CREATE SCHEMA public;

CREATE TABLE lineitem (
    l_orderkey       integer NOT NULL,
    l_partkey        integer NOT NULL,
    l_suppkey        integer NOT NULL,
    l_linenumber     integer NOT NULL,
    l_quantity       decimal(15, 2) NOT NULL,
    l_extendedprice  decimal(15, 2) NOT NULL,
    l_discount       decimal(15, 2) NOT NULL,
    l_tax            decimal(15, 2) NOT NULL,
    l_returnflag     text NOT NULL,
    l_linestatus     text NOT NULL,
    l_shipdate       date NOT NULL,
    l_commitdate     date NOT NULL,
    l_receiptdate    date NOT NULL,
    l_shipinstruct   text NOT NULL,
    l_shipmode       text NOT NULL,
    l_comment        text NOT NULL
);

CREATE INDEX pk_lineitem_orderkey_linenumber ON lineitem (l_orderkey, l_linenumber);
CREATE INDEX fk_lineitem_orderkey ON lineitem (l_orderkey ASC);

CREATE TABLE orders (
    o_orderkey       integer,
    o_custkey        integer NOT NULL,
    o_orderstatus    text NOT NULL,
    o_totalprice     decimal(15, 2) NOT NULL,
    o_orderdate      DATE NOT NULL,
    o_orderpriority  text NOT NULL,
    o_clerk          text NOT NULL,
    o_shippriority   integer NOT NULL,
    o_comment        text NOT NULL
);

CREATE INDEX pk_orders_orderkey ON orders (o_orderkey);
CREATE INDEX fk_orders_custkey ON orders (o_custkey ASC);

SELECT MAX( o_orderkey ) FROM lineitem
JOIN orders ON ( l_orderkey = o_orderkey )
WHERE l_linenumber = 1 AND o_custkey = 196 AND l_commitDATE = '1994-09-23';```

Reply:
@wangandi  I am afraid this prevents further automated testing -- basically a large space of queries that are not cross joins and have an actual join condition will panic. I will be able to continue after it is moved out of the way.

```get_relation_with_reduction_pushed``` is mentioned multiple times in the stack trace, so the line number of this comment may not be the most relevant one.

Reply:
Thanks for finding the crash! It should be fixed now!